File: python/src/taskulus/maintenance.py
"""Maintenance command implementations."""

from __future__ import annotations

import json
from pathlib import Path
from dataclasses import dataclass
from typing import Dict, List, Optional, Set

from pydantic import ValidationError as PydanticValidationError

from taskulus.config_loader import ConfigurationError, load_project_configuration
from taskulus.dependencies import ALLOWED_DEPENDENCY_TYPES
from taskulus.hierarchy import InvalidHierarchyError, validate_parent_child_relationship
from taskulus.models import IssueData, ProjectConfiguration
from taskulus.project import (
    ProjectMarkerError,
    get_configuration_path,
    load_project_directory,
)
from taskulus.workflows import get_workflow_for_issue_type


class ProjectValidationError(RuntimeError):
    """Raised when project validation fails."""


class ProjectStatsError(RuntimeError):
    """Raised when project statistics cannot be computed."""


@dataclass(frozen=True)
class ProjectStats:
    """Aggregate issue statistics for a project.

    :param total: Total number of issues.
    :type total: int
    :param open_count: Number of open issues.
    :type open_count: int
    :param closed_count: Number of closed issues.
    :type closed_count: int
    :param type_counts: Counts by issue type.
    :type type_counts: Dict[str, int]
    """

    total: int
    open_count: int
    closed_count: int
    type_counts: Dict[str, int]


def validate_project(root: Path) -> None:
    """Validate issue data and configuration for a Taskulus project.

    :param root: Repository root path.
    :type root: Path
    :raises ProjectValidationError: If validation fails.
    """
    try:
        project_dir = load_project_directory(root)
    except ProjectMarkerError as error:
        raise ProjectValidationError(str(error)) from error

    issues_dir = project_dir / "issues"
    if not issues_dir.exists():
        raise ProjectValidationError("issues directory missing")

    try:
        configuration = load_project_configuration(get_configuration_path(project_dir))
    except ConfigurationError as error:
        raise ProjectValidationError(str(error)) from error

    errors: List[str] = []
    issues: Dict[str, IssueData] = {}
    for issue_path in sorted(issues_dir.glob("*.json"), key=lambda path: path.name):
        issue = _load_issue(issue_path, errors)
        if issue is None:
            continue
        if issue.identifier in issues:
            errors.append(f"{issue_path.name}: duplicate issue id '{issue.identifier}'")
            continue
        _validate_issue_fields(issue_path.name, issue, configuration, errors)
        issues[issue.identifier] = issue

    _validate_references(issues, configuration, errors)

    if errors:
        raise ProjectValidationError(_format_errors(errors))


def collect_project_stats(root: Path) -> ProjectStats:
    """Collect project statistics from issue data.

    :param root: Repository root path.
    :type root: Path
    :return: Aggregated project statistics.
    :rtype: ProjectStats
    :raises ProjectStatsError: If stats cannot be computed.
    """
    try:
        project_dir = load_project_directory(root)
    except ProjectMarkerError as error:
        raise ProjectStatsError(str(error)) from error

    issues_dir = project_dir / "issues"
    if not issues_dir.exists():
        raise ProjectStatsError("issues directory missing")

    issues: List[IssueData] = []
    for issue_path in sorted(issues_dir.glob("*.json"), key=lambda path: path.name):
        try:
            payload = json.loads(issue_path.read_text(encoding="utf-8"))
        except json.JSONDecodeError as error:
            raise ProjectStatsError(
                f"{issue_path.name}: invalid json: {error}"
            ) from error
        try:
            issues.append(IssueData.model_validate(payload))
        except PydanticValidationError as error:
            raise ProjectStatsError(
                f"{issue_path.name}: invalid issue data: {error}"
            ) from error

    total = len(issues)
    closed_count = sum(1 for issue in issues if issue.status == "closed")
    open_count = total - closed_count
    type_counts: Dict[str, int] = {}
    for issue in issues:
        type_counts[issue.issue_type] = type_counts.get(issue.issue_type, 0) + 1

    return ProjectStats(
        total=total,
        open_count=open_count,
        closed_count=closed_count,
        type_counts=type_counts,
    )


def _load_issue(issue_path: Path, errors: List[str]) -> Optional[IssueData]:
    try:
        payload = json.loads(issue_path.read_text(encoding="utf-8"))
    except OSError as error:
        errors.append(f"{issue_path.name}: unable to read issue: {error}")
        return None
    except json.JSONDecodeError as error:
        errors.append(f"{issue_path.name}: invalid json: {error}")
        return None
    try:
        return IssueData.model_validate(payload)
    except PydanticValidationError as error:
        errors.append(f"{issue_path.name}: invalid issue data: {error}")
        return None


def _validate_issue_fields(
    filename: str,
    issue: IssueData,
    configuration: ProjectConfiguration,
    errors: List[str],
) -> None:
    valid_types = configuration.hierarchy + configuration.types
    if issue.identifier != Path(filename).stem:
        errors.append(
            f"{filename}: issue id '{issue.identifier}' does not match filename"
        )
    if issue.issue_type not in valid_types:
        errors.append(f"{filename}: unknown issue type '{issue.issue_type}'")
    if issue.priority not in configuration.priorities:
        errors.append(f"{filename}: invalid priority '{issue.priority}'")

    statuses = _collect_workflow_statuses(configuration, issue.issue_type, errors)
    if statuses is not None and issue.status not in statuses:
        errors.append(f"{filename}: invalid status '{issue.status}'")

    if issue.status == "closed" and issue.closed_at is None:
        errors.append(f"{filename}: closed issues must have closed_at set")
    if issue.status != "closed" and issue.closed_at is not None:
        errors.append(f"{filename}: non-closed issues must not set closed_at")

    for dependency in issue.dependencies:
        if dependency.dependency_type not in ALLOWED_DEPENDENCY_TYPES:
            errors.append(
                f"{filename}: invalid dependency type '{dependency.dependency_type}'"
            )


def _collect_workflow_statuses(
    configuration: ProjectConfiguration,
    issue_type: str,
    errors: List[str],
) -> Optional[Set[str]]:
    try:
        workflow = get_workflow_for_issue_type(configuration, issue_type)
    except ValueError as error:
        errors.append(str(error))
        return None
    statuses: Set[str] = set(workflow.keys())
    for transitions in workflow.values():
        statuses.update(transitions)
    return statuses


def _validate_references(
    issues: Dict[str, IssueData],
    configuration: ProjectConfiguration,
    errors: List[str],
) -> None:
    for issue in issues.values():
        if issue.parent:
            parent = issues.get(issue.parent)
            if parent is None:
                errors.append(
                    f"{issue.identifier}: parent '{issue.parent}' does not exist"
                )
            else:
                try:
                    validate_parent_child_relationship(
                        configuration, parent.issue_type, issue.issue_type
                    )
                except InvalidHierarchyError as error:
                    errors.append(f"{issue.identifier}: {error}")

        for dependency in issue.dependencies:
            if dependency.target not in issues:
                errors.append(
                    f"{issue.identifier}: dependency target '{dependency.target}' does not exist"
                )


def _format_errors(errors: List[str]) -> str:
    return "validation failed:\n" + "\n".join(errors)

File: rust/src/maintenance.rs
//! Maintenance command implementations.

use std::collections::{BTreeMap, BTreeSet};
use std::fs;
use std::path::Path;

use crate::config_loader::load_project_configuration;
use crate::error::TaskulusError;
use crate::file_io::{get_configuration_path, load_project_directory};
use crate::hierarchy::validate_parent_child_relationship;
use crate::models::IssueData;
use crate::workflows::get_workflow_for_issue_type;

const ALLOWED_DEPENDENCY_TYPES: [&str; 2] = ["blocked-by", "relates-to"];

/// Aggregate issue statistics for a project.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ProjectStats {
    pub total: usize,
    pub open_count: usize,
    pub closed_count: usize,
    pub type_counts: BTreeMap<String, usize>,
}

/// Validate issue data and configuration for a Taskulus project.
///
/// # Arguments
/// * `root` - Repository root path.
///
/// # Errors
/// Returns `TaskulusError::IssueOperation` if validation fails.
pub fn validate_project(root: &Path) -> Result<(), TaskulusError> {
    let project_dir = load_project_directory(root)?;
    let issues_dir = project_dir.join("issues");
    if !issues_dir.exists() {
        return Err(TaskulusError::IssueOperation(
            "issues directory missing".to_string(),
        ));
    }

    let configuration =
        load_project_configuration(&get_configuration_path(project_dir.as_path())?)?;

    let mut errors: Vec<String> = Vec::new();
    let mut issues: BTreeMap<String, IssueData> = BTreeMap::new();

    let mut paths: Vec<_> = fs::read_dir(&issues_dir)
        .map_err(|error| TaskulusError::Io(error.to_string()))?
        .filter_map(|entry| entry.ok())
        .map(|entry| entry.path())
        .filter(|path| path.extension().and_then(|ext| ext.to_str()) == Some("json"))
        .collect();
    paths.sort_by(|left, right| left.file_name().cmp(&right.file_name()));

    for path in paths {
        let filename = path
            .file_name()
            .and_then(|name| name.to_str())
            .unwrap_or("unknown");
        let contents = match fs::read_to_string(&path) {
            Ok(contents) => contents,
            Err(error) => {
                errors.push(format!("{filename}: unable to read issue: {error}"));
                continue;
            }
        };

        let payload: serde_json::Value = match serde_json::from_str(&contents) {
            Ok(payload) => payload,
            Err(error) => {
                errors.push(format!("{filename}: invalid json: {error}"));
                continue;
            }
        };

        let issue: IssueData = match serde_json::from_value(payload) {
            Ok(issue) => issue,
            Err(error) => {
                errors.push(format!("{filename}: invalid issue data: {error}"));
                continue;
            }
        };

        if issues.contains_key(&issue.identifier) {
            errors.push(format!(
                "{filename}: duplicate issue id '{}'",
                issue.identifier
            ));
            continue;
        }

        validate_issue_fields(filename, &issue, &configuration, &mut errors);
        issues.insert(issue.identifier.clone(), issue);
    }

    validate_references(&issues, &configuration, &mut errors);

    if errors.is_empty() {
        Ok(())
    } else {
        Err(TaskulusError::IssueOperation(format_errors(&errors)))
    }
}

/// Collect project statistics from issue data.
///
/// # Arguments
/// * `root` - Repository root path.
///
/// # Returns
/// Aggregated project statistics.
///
/// # Errors
/// Returns `TaskulusError::IssueOperation` if stats cannot be computed.
pub fn collect_project_stats(root: &Path) -> Result<ProjectStats, TaskulusError> {
    let project_dir = load_project_directory(root)?;
    let issues_dir = project_dir.join("issues");
    if !issues_dir.exists() {
        return Err(TaskulusError::IssueOperation(
            "issues directory missing".to_string(),
        ));
    }

    let mut issues: Vec<IssueData> = Vec::new();
    for entry in fs::read_dir(&issues_dir).map_err(|error| TaskulusError::Io(error.to_string()))? {
        let entry = entry.map_err(|error| TaskulusError::Io(error.to_string()))?;
        let path = entry.path();
        if path.extension().and_then(|ext| ext.to_str()) != Some("json") {
            continue;
        }
        let contents =
            fs::read_to_string(&path).map_err(|error| TaskulusError::Io(error.to_string()))?;
        let filename = path
            .file_name()
            .and_then(|name| name.to_str())
            .unwrap_or("unknown");
        let payload: serde_json::Value = serde_json::from_str(&contents).map_err(|error| {
            TaskulusError::IssueOperation(format!("{filename}: invalid json: {error}"))
        })?;
        let issue: IssueData = serde_json::from_value(payload).map_err(|error| {
            TaskulusError::IssueOperation(format!("{filename}: invalid issue data: {error}"))
        })?;
        issues.push(issue);
    }

    let total = issues.len();
    let closed_count = issues
        .iter()
        .filter(|issue| issue.status == "closed")
        .count();
    let open_count = total - closed_count;
    let mut type_counts: BTreeMap<String, usize> = BTreeMap::new();
    for issue in issues {
        *type_counts.entry(issue.issue_type).or_insert(0) += 1;
    }

    Ok(ProjectStats {
        total,
        open_count,
        closed_count,
        type_counts,
    })
}

fn validate_issue_fields(
    filename: &str,
    issue: &IssueData,
    configuration: &crate::models::ProjectConfiguration,
    errors: &mut Vec<String>,
) {
    let expected_id = Path::new(filename)
        .file_stem()
        .and_then(|stem| stem.to_str())
        .unwrap_or(filename);
    if issue.identifier != expected_id {
        errors.push(format!(
            "{filename}: issue id '{}' does not match filename",
            issue.identifier
        ));
    }

    let mut valid_types = configuration.hierarchy.clone();
    valid_types.extend(configuration.types.clone());
    if !valid_types.iter().any(|entry| entry == &issue.issue_type) {
        errors.push(format!(
            "{filename}: unknown issue type '{}'",
            issue.issue_type
        ));
    }

    let priority_value = if (0..=u8::MAX as i32).contains(&issue.priority) {
        Some(issue.priority as u8)
    } else {
        None
    };
    let priority_valid =
        priority_value.is_some_and(|value| configuration.priorities.contains_key(&value));
    if !priority_valid {
        errors.push(format!("{filename}: invalid priority '{}'", issue.priority));
    }

    if let Ok(statuses) = collect_workflow_statuses(configuration, &issue.issue_type) {
        if !statuses.contains(&issue.status) {
            errors.push(format!("{filename}: invalid status '{}'", issue.status));
        }
    }

    if issue.status == "closed" && issue.closed_at.is_none() {
        errors.push(format!("{filename}: closed issues must have closed_at set"));
    }
    if issue.status != "closed" && issue.closed_at.is_some() {
        errors.push(format!(
            "{filename}: non-closed issues must not set closed_at"
        ));
    }

    for dependency in &issue.dependencies {
        if !ALLOWED_DEPENDENCY_TYPES
            .iter()
            .any(|entry| *entry == dependency.dependency_type)
        {
            errors.push(format!(
                "{filename}: invalid dependency type '{}'",
                dependency.dependency_type
            ));
        }
    }
}

fn collect_workflow_statuses(
    configuration: &crate::models::ProjectConfiguration,
    issue_type: &str,
) -> Result<BTreeSet<String>, TaskulusError> {
    let workflow = get_workflow_for_issue_type(configuration, issue_type)?;
    let mut statuses: BTreeSet<String> = workflow.keys().cloned().collect();
    for transitions in workflow.values() {
        statuses.extend(transitions.iter().cloned());
    }
    Ok(statuses)
}

fn validate_references(
    issues: &BTreeMap<String, IssueData>,
    configuration: &crate::models::ProjectConfiguration,
    errors: &mut Vec<String>,
) {
    for issue in issues.values() {
        if let Some(parent_id) = &issue.parent {
            match issues.get(parent_id) {
                Some(parent_issue) => {
                    if let Err(error) = validate_parent_child_relationship(
                        configuration,
                        &parent_issue.issue_type,
                        &issue.issue_type,
                    ) {
                        errors.push(format!("{}: {}", issue.identifier, error));
                    }
                }
                None => errors.push(format!(
                    "{}: parent '{}' does not exist",
                    issue.identifier, parent_id
                )),
            }
        }

        for dependency in &issue.dependencies {
            if !issues.contains_key(&dependency.target) {
                errors.push(format!(
                    "{}: dependency target '{}' does not exist",
                    issue.identifier, dependency.target
                ));
            }
        }
    }
}

fn format_errors(errors: &[String]) -> String {
    format!("validation failed:\n{}", errors.join("\n"))
}
